### PR TITLE
Sidebar more css classes and styles

### LINKED ISSUE #102: extension: sidebar issues on certain sites (e.g. truncation/scrolling/etc)
Feel free to comment here if you find more such examples

- [ ] https://arthurperret.fr/2020/05/22/ecrire-et-editer/ (@bepolymathe)
      same for Hypothesis, so potentially site markup is somehow unusual
- [ ] https://www.quaternum.net (@bepolymathe)
      possibly something to do with Jekyll?
- [ ] https://psychonautwiki.org/wiki/Cannabidiol (sidebar detaches when scrolled)
- [ ] https://www.bloomberg.com/europe (might be behind popups) https://github.com/karlicoss/promnesia/issues/146#issue-727850381

### PR DESCRIPTION
**Pull request by yoyurec**
* Fixes https://github.com/karlicoss/promnesia/issues/102
* Added more classes to markup for deeper restyling possibilities
* Header "sticky", only visits scrolled
* Added '.button' styles
* Main buttons more recognisible - with emoji icons
* Added badges counter for sources filter
* Added `active` status for "Mark visited" button
* Added `active` status for selected source filter (!)
* Reworked a lot of styling for `#visits` elements
   * bookmark source items with icon ⭐
   * box colors
   * timestamp monospaced font
 * Text/bg/border colors use CSS variables

Before:   

After:
image

### CODE REVIEW

File: extension/webpack.config.js
Comment (karlicoss) [inline extension/webpack.config.js]: we should leave it out, would be nice to figure out why the ./build script not working. Might result in some nasty bugs in tests etc..

File: extension/src/options.js
Comment (karlicoss) [inline extension/src/options.js]: should be wrapped in a comment?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: I'd rather avoid loading random fonts as the default (esp from Google) -- let's leave it up to the user settings :)

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: best to name it .src-count (or scr-total).
'number' usually refers to the index rather than total count

File: extension/src/sidebar.js
Comment (karlicoss) [inline extension/src/sidebar.js]: we should use `safeSetInnerHTML` function (otherwise it might trip up some Mozilla security checks)

File: extension/src/sidebar-outer.css
Comment (karlicoss) [inline extension/src/sidebar-outer.css:41]: hmm, is this max int? is there a reason to set it like that?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: IMO it's way too subtle -- maybe should set some thickness as well?

File: extension/src/sidebar.js
Comment (karlicoss) [inline extension/src/sidebar.js]: maybe instead of classes, use button-close/button-search etc as `id`? since they are unique elements

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: image
hmm I think this padding makes it misaligned with the other buttons -- is it necessary?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm, not sure what it's supposed to do, but it doesnt look like it capitalizes anything for me

File: extension/src/sidebar.js
Comment (karlicoss) [inline extension/src/sidebar.js:183]: TODO container id changed?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: imo default font is fine, let's just keep it up to the configuration?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm, do we need word-wrap if we use nowrap above?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css:239]: hmm, why? I get there is probably little point to select text there, but I'd prefer to keep css as small as possible :)

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css:238]: hm, why? doesn't seem to have any effect for me

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: what's that for? doesn't seem to impact anything for me

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: box-sizing/box-shadow don't seem to impact anything for me?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: I guess it seems useful, just wondering how did you notice that it's necessary? i.e. seems to only have effect on the phone?

File: extension/src/search.html
Comment (karlicoss) [inline extension/src/search.html:11]: hmm, why remove type=submit?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm, are display: flex and flex-flow necessary? it behaves the same without them

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm, what's it for here?
I usually try to write comments for obscure css properties so it's easier for people to understand why it's here, so it's not lost during refactoring

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm, why split it into margin and padding?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm, what does it do?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: similar to above, renameto --source-count-color?

File: extension/src/options.js
Comment (karlicoss) [inline extension/src/options.js]: nit: typos -- tweak/light

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: imo kinda subjective, so maybe keep in the private config? but don't mind it either

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm, these names are really not descriptive, can we maybe add an explanation/link to the screenshot explaining what are these? 
(fine to do later once we settle on the final version)

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: hmm imo it's subjective -- shall we keep it in your private config?

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: isn't it kinda pointless to have 1em as the font size? it doesn't do anything right?

File: extension/src/sidebar-outer.css
Comment (karlicoss) [inline extension/src/sidebar-outer.css:8]: hmm what's it for?

File: extension/src/options.js
Comment (yoyurec) [reply extension/src/options.js]: No, sidebar uses it )
Importing font implemented in iframe scope (as it should), but some sites (for. ex. github) have warnings, fonts not loaded:
`Refused to load the stylesheet 'https://fonts.googleapis.com/css2?family=Fira+Code:wght@400;600&family=PT+Serif:ital,wght@0,400;0,700;1,400;1,700&display=swap' because it violates the following Content Security Policy directive: "style-src 'unsafe-inline' github.githubassets.com". Note that 'style-src-elem' was not explicitly set, so 'style-src' is used as a fallback.`

So font duplicated in options, b'cose their css scope = initial page scope, no policy warnings.

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: ok, will remove it and put to my own options settings )
anyway it looks almost the same, but little narrower

File: extension/src/sidebar-outer.css
Comment (karlicoss) [inline extension/src/sidebar-outer.css:3]: oh now I see why it's here https://github.com/karlicoss/promnesia/issues/102#issuecomment-1114237352
can you add a link to the issue please? because otherwise it's not obvious what this code is for :)

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: ok ) 
hm... initally i want to name it `src-count` 🤔

File: extension/src/sidebar.js
Comment (yoyurec) [reply extension/src/sidebar.js]: smart!

File: extension/webpack.config.js
Comment (yoyurec) [reply extension/webpack.config.js]: you mean delete?

File: extension/src/sidebar-outer.css
Comment (yoyurec) [reply extension/src/sidebar-outer.css:41]: yes, this is max z-index. will beat all popups, cookie blocks, `scroll to top` arrows, etc...
reason is clear and simple - some selements bleads throung sidebar ((
reason same as before, but 9999 isn't big enougth

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: give me the number )

File: extension/src/sidebar.js
Comment (yoyurec) [reply extension/src/sidebar.js]: ok

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: will make it beautiful

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: it capitalize word `filter: ` (BTW, why not "Filter:" in initial js?) + source buttons....
can you put screen?

File: extension/src/sidebar.js
Comment (yoyurec) [reply extension/src/sidebar.js:183]: TODO? how to see it?

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: default font have no proper emoji icons ((
windows:
image

File: extension/src/sidebar-outer.css
Comment (karlicoss) [reply extension/src/sidebar-outer.css:41]: aah gotcha -- didn't expect there would be any pages with such z-inde

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css:239]: common button ui behavior. sometimes users do clicking and text selects. but it shouldn't

File: extension/src/sidebar.js
Comment (karlicoss) [reply extension/src/sidebar.js:183]: ah sorry this was todo for me -- forgot to remove it, this bit is good :)

File: extension/webpack.config.js
Comment (karlicoss) [reply extension/webpack.config.js]: yeah, you could keep it locally but not commit?

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css:238]: also common buttons stuff.
removing default platrform styling...
https://css-tricks.com/almanac/properties/a/appearance/
something like reset.css... maybe mostly for mobiles...

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: dunno. copypasted from button generator. meybe used in some complicated content case
will delete

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: shadows impacts, look closer ))

box-sizing can be removed.

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: can delete. code from generator

File: extension/src/search.html
Comment (yoyurec) [reply extension/src/search.html:11]: is it removed?

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: default is row
image

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: when init page have scroll. and our sidebar have scroll. when you scroll with mouse wheel inside sidebar and reach its end, but countinue to scroll with mouse wheel - whole init page will countinue to scroll. it's by default and it's annoying. 
` overscroll-behavior: contain;` disables it.

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: for better look in case of adding border (was in my previous revisions). will delete

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: from generator. will delete.

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: right

File: extension/src/options.js
Comment (yoyurec) [reply extension/src/options.js]: tnx. it was deep **niht** 😴

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: ok

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: yes

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: for default font it's not need, is i see...

File: extension/src/sidebar-outer.css
Comment (yoyurec) [reply extension/src/sidebar-outer.css:8]: for issue https://github.com/karlicoss/promnesia/issues/102

File: extension/src/sidebar-outer.css
Comment (yoyurec) [reply extension/src/sidebar-outer.css:3]: ok

File: extension/src/search.html
Comment (karlicoss) [reply extension/src/search.html:11]: :man_facepalming:  sorry, not sure where I looked

File: extension/webpack.config.js
Comment (yoyurec) [reply extension/webpack.config.js]: ok

File: extension/src/sidebar.css
Comment (karlicoss) [reply extension/src/sidebar.css]: yep, but is `display: flex` necessary at all?

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: it's different )
dunno. will delete both

Comment (karlicoss) [reply extension/src/sidebar.css]: the `white-space: nowrap` is definitely useful, you can see it if you have a long tag with whitespace, and try a narrower screen in the devtools. Just not sure about the break-word :)

File: extension/src/sidebar.css
Comment (karlicoss) [reply extension/src/sidebar.css]: ah, fair enough -- can you add a comment about this please :)

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: it's "2nd layer" of variables ))
most important selectors has own good named vars:
image

but inside them i also using vars... THE THEME vars, their names here a little not speackable.
they reused in different properties, so can't be named `---border-color` as it's also used in shadow, etc...

it's like abstract vars, and then below more linked to blocks..

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: done

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: > IMO it's way too subtle

you can use your own options config ahahahahahah )))

Comment (yoyurec) [reply extension/src/sidebar.css]: i tried 2px. it's tooooo big and ugly (((

File: extension/src/sidebar.css
Comment (karlicoss) [reply extension/src/sidebar.css]: ah -- I think it didn't work for some reason, but seem that you removed it now anyway
Rither way, imo if the user wanted to capitalize source names they would just name them from the capital letter. Also it makes it a bit misleading for debugging etc since the source names in data would be different from what they appear as.
renaming to `Filter: ` is fine though, I just have a very relaxed attitude towards capital letters xD

File: extension/src/sidebar.css
Comment (karlicoss) [reply extension/src/sidebar.css]: ok! Not a big deal, so fine to keep as it is if you prefer :) 
could also use coloring change or something like that, but I don't mind to keep it as is

File: extension/src/sidebar.css
Comment (karlicoss) [inline extension/src/sidebar.css]: typo :)

File: extension/src/sidebar.css
Comment (karlicoss) [reply extension/src/sidebar.css]: hm, okay makes sense

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: like that?
image

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: at first i  want to name them `ligh`, `lighter`, etc... but then i introduced dark mode and in became little confusing.
will try something like `primary`, `secondary`... or....
at least sinplify (again and again 😁) amount of used colors more...

File: extension/src/sidebar.css
Comment (karlicoss) [reply extension/src/sidebar.css]: yup, looks good!

File: extension/src/sidebar.js
Comment (yoyurec) [inline extension/src/sidebar.js]: em... it doesn't clean ((
filters populated `onCompleted`, i need to clean them. previos code did it well...
image

Comment (karlicoss) [reply extension/src/sidebar.js]: ah, that's odd! I think we should add cleaning to `safeSetInnerHTML`, must have missed it by accident

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: done!

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: tnx

File: extension/src/sidebar.js
Comment (yoyurec) [reply extension/src/sidebar.js]: i think we can use native API https://developer.mozilla.org/en-US/docs/Web/API/Element/replaceChildren

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: yes, it allow to have scroll for #visits. 
overflow-y: auto; + height:100% not working correct.

File: extension/src/sidebar.css
Comment (yoyurec) [reply extension/src/sidebar.css]: i'll make it lighter, but it do visually beautify a lot )

Comment (karlicoss) [reply extension/src/sidebar.css]: yeah it's fine to keep here, don't worry

File: extension/src/sidebar.js
Comment (karlicoss) [reply extension/src/sidebar.js]: looks like JS flow is complaining, but I think it's just because `replaceChildren` was added relatively recently, in 2020
https://developer.mozilla.org/en-US/docs/Web/API/Element#browser_compatibility
I think it should be fine just to suppress its error in this case, 2020 is quite old in browser versions
Or even better maybe wrap it in try/catch just in case (with a comment). Worst case the extension will be slighly broken for the user, but not completely :)

**Comment by yoyurec** [comment]
phew... looks like all comments + few new spotted things fixed!

File: extension/src/sidebar.js
Comment (karlicoss) [inline extension/src/sidebar.js]: code seems misaligned in this block, can you shift it one space to the left?

File: extension/webpack.config.js
Comment (karlicoss) [reply extension/webpack.config.js]: hopefully just this left :)

**Review by karlicoss — CHANGES_REQUESTED** [review]

File: extension/webpack.config.js
Comment (yoyurec) [reply extension/webpack.config.js]: gone )

File: extension/src/sidebar.js
Comment (yoyurec) [reply extension/src/sidebar.js]: done

**Review by karlicoss — APPROVED** [review]
awesome, thanks so much! :tada: 
will merge once CI completes

**Pull request merged**
